### PR TITLE
Implement semVerLevel query parameter (this is SemVer 2.0.0 protocol support)

### PR DESCRIPTION
**Pull request by joelverhagen**
This cherry-picks the PR https://github.com/NuGet/NuGet.Server/pull/27 from `master` into `dev`. Since `dev` and `master` have diverged significantly, it was a very manual process. There is a lot more testing in `dev` so this PR matches that.

Additionally, this PR introduces an improvement not elegantly done in WCF: mutate OData self/identity URLs. A SemVer 2.0.0 version can contain a plus so package self/identity URLs should have version metadata stripped out (e.g. `1.0.0+git` becomes `1.0.0`). This is because IIS rejects pluses in URLs by default. I could not find a good way to get this working in `master` (WCF) so I introduced it here in `dev` (Web API).

The design document is here:
[SemVer 2.0.0 Protocol](https://github.com/NuGet/Home/wiki/Semver-2.0.0-Protocol)

/cc @emgarten @dtivel @maartenba

### CODE REVIEW

File: src/NuGet.Server.Core/Infrastructure/ServerPackageStore.cs [hunk 1]
```diff
@@ -103,17 +103,16 @@ public void Remove(string id, SemanticVersion version, bool enableDelisting)
...
             {
                 allPackages.Add(serverPackage);
             }
         }
 
-        private ServerPackage CreateServerPackage(
+        private ServerPackage CreateServerPackageOrNull(
             IPackage package,
-            bool enableDelisting,
-            CancellationToken token)
```
Comment (joelverhagen) [inline src/NuGet.Server.Core/Infrastructure/ServerPackageStore.cs]: This `token` removal addresses this follow-up comment: https://github.com/NuGet/NuGet.Server/pull/28#discussion_r95200934.

File: src/NuGet.Server.Core/Infrastructure/IServerPackageRepository.cs [hunk 2]
```diff
@@ -13,12 +13,13 @@ public interface IServerPackageRepository
 
         Task AddPackageAsync(IPackage package, CancellationToken token);
 
-        Task<IEnumerable<IServerPackage>> GetPackagesAsync(CancellationToken token);
+        Task<IEnumerable<IServerPackage>> GetPackagesAsync(ClientCompatibility compatiblity, CancellationToken token);
```
Comment (dtivel) [inline src/NuGet.Server.Core/Infrastructure/IServerPackageRepository.cs]: compatibility

Comment (joelverhagen) [inline src/NuGet.Server.Core/Infrastructure/IServerPackageRepository.cs]: Fixed (+ one other place)

File: test/NuGet.Server.V2.Tests/Infrastructure/ControllerTestHelpers.cs [hunk 3]
```diff
@@ -98,6 +100,45 @@ public static Mock<IServerPackageRepository> SetupTestPackageRepository()
                 },
                 new ServerPackage
                 {
+                    //PackageRegistration = barPackage,
```
Comment (dtivel) [inline test/NuGet.Server.V2.Tests/Infrastructure/ControllerTestHelpers.cs]: Why keep the commented out lines?

File: test/NuGet.Server.Core.Tests/ClientCompatibilityFactoryTests.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+            // Arrange & Act
+            var actual = ClientCompatibilityFactory.FromProperties(semVerLevel);
+
+            // Assert
+            Assert.Equal(expected, actual.SemVerLevel);
+        }
+
+        [Fact]
+        public void FromProperties_AllowsNullSemVerLevel()
```
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/ClientCompatibilityFactoryTests.cs]: Good point!

File: test/NuGet.Server.V2.Tests/Infrastructure/ControllerTestHelpers.cs (hunk 3 above)
Comment (joelverhagen) [inline test/NuGet.Server.V2.Tests/Infrastructure/ControllerTestHelpers.cs]: Pre-existing. Will remove.

File: test/NuGet.Server.Core.Tests/ClientCompatibilityFactoryTests.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        [Theory]
+        [InlineData(null, "1.0.0")]
+        [InlineData("0", "1.0.0")]
+        [InlineData("a", "1.0.0")]
+        [InlineData("0.0.0", "0.0.0")]
+        [InlineData("1.0.0", "1.0.0")]
+        [InlineData("2.0.0", "2.0.0")]
+        [InlineData("2.0.0-rc.1", "2.0.0-rc.1")]
+        [InlineData("3.0.0", "3.0.0")]
+        public void FromProperties_SetsSemVerLevel(string semVerLevel, string unparsedExpected)
```
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/ClientCompatibilityFactoryTests.cs]: Fixed.

File: test/NuGet.Server.Core.Tests/ClientCompatibilityFactoryTests.cs (hunk 4 above)
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/ClientCompatibilityFactoryTests.cs]: Fixed.

File: test/NuGet.Server.Core.Tests/PackageExtensionsTest.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        [Theory]
+        [InlineData(true, true, false, false, 1, true, true)]
+        [InlineData(false, false, true, true, 1, false, false)]
+        [InlineData(true, false, false, false, 1, true, false)]
+        [InlineData(false, false, true, true, 2, true, true)]
+        [InlineData(true, true, false, false, 2, false, false)]
+        [InlineData(false, false, true, false, 2, true, false)]
+        public void AsODataPackage_PicksCorrectLatestProperties(
+            bool v1AbsLatest, bool v1Latest, bool v2AbsLatest, bool v2Latest, int level, bool absLatest, bool latest)
```
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/PackageExtensionsTest.cs]: Fixed.

File: test/NuGet.Server.Core.Tests/PackageExtensionsTest.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                Version = new SemanticVersion("0.1.0-rc.1"),
+                Authors = Enumerable.Empty<string>(),
+                Owners = Enumerable.Empty<string>(),
+
+                SemVer1IsAbsoluteLatest = v1AbsLatest,
+                SemVer1IsLatest = v1Latest,
+                SemVer2IsAbsoluteLatest = v2AbsLatest,
+                SemVer2IsLatest = v2Latest,
+            };
+            var semVerLevel = new SemanticVersion(level, 0, 0, 0);
```
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/PackageExtensionsTest.cs]: Fixed.

File: test/NuGet.Server.V2.Tests/Infrastructure/ControllerTestHelpers.cs (hunk 3 above)
Comment (joelverhagen) [inline test/NuGet.Server.V2.Tests/Infrastructure/ControllerTestHelpers.cs]: Fixed.

File: src/NuGet.Server.Core/Infrastructure/ServerPackageRepository.cs [hunk 8]
```diff
@@ -119,7 +119,9 @@ private ServerPackageCache InitializeServerPackageStore()
         /// Package cache containing packages metadata. 
         /// This data is generated if it does not exist already.
         /// </summary>
-        public async Task<IEnumerable<IServerPackage>> GetPackagesAsync(CancellationToken token)
+        public async Task<IEnumerable<IServerPackage>> GetPackagesAsync(
+            ClientCompatibility compatiblity,
```
Comment (joelverhagen) [inline src/NuGet.Server.Core/Infrastructure/ServerPackageRepository.cs]: Fixed.

**Pull request merged**